docs(v4): new router provider (#3689)
refinedev/refine: documentation/docs/packages/documentation/routers/nextjs-router.md

---
id: nextjs
title: Next.js
---

> **refine** provides router bindings and utiliities for [Next.js](https://nextjs.org/). This package will provide easy integration between **refine** and **Next.js** for both existing projects and new projects without giving up the benefits of **Next.js**.

Reviewer: ```suggestion
**refine** offers router bindings and utilities that are specifically designed for Next.js, providing a seamless integration experience for both new and existing projects.

By using Refine with Next.js, developers can take advantage of Next.js's benefits, such as server-side rendering and client-side routing, while also utilizing the additional features and functionality that Refine provides.
```

```bash
npm i @pankod/refine-nextjs-router
```

[Refer to the Router Provider documentation for detailed information. &#8594][routerprovider]

:::note Legacy Router

> `@pankod/refine-nextjs-router` also exports the legacy router provider and it will be available until the next major version of **refine**. It is recommended to use the new router provider instead of the legacy one.
> 
> If you are using the legacy router provider, it can be imported from `@pankod/refine-nextjs-router/legacy` for the `/pages` directory and `@pankod/refine-nextjs-router/legacy-app` for the `/app` directory and passed to the `legacyRouterProvider` prop of the `Refine` component.

Reviewer: ```suggestion
The legacy router provider of v3 is also exported by @pankod/refine-nextjs-router, and will continue to be available until the next major version of **refine** is released. However, it is recommended that developers use the new router provider instead of the legacy one.

If you still want to use the legacy router provider, it can be imported from `@pankod/refine-nextjs-router/legacy` for the `/pages` directory, and `@pankod/refine-nextjs-router/legacy-app` for the `/app` directory.

The legacy router provider should then be passed to the legacyRouterProvider prop of the Refine component.
```

:::

## Basic Usage

### `/pages` Directory

> We'll use the `routerProvider` from `@pankod/refine-nextjs-router` to set up the router bindings for **refine**. We'll define the action routes for our resources in the `resources` array and define our pages in the `pages` directory.
> 
> We'll create four pages for our resources:
> 
> - `pages/posts/index.tsx` - List page for posts
> - `pages/posts/show/[id].tsx` - Detail page for posts
> - `pages/categories/index.tsx` - List page for categories
> - `pages/categories/show/[id].tsx` - Detail page for categories
> 
> And we'll create one page for the index route and use it to redirect to the `posts` resource:
> 
> - `pages/index.tsx` - Index page
> 
> Let's start with the initialization of the **refine** app in the `_app.tsx` file:

Reviewer: ```suggestion
To set up the router bindings for **refine**, we will utilize the `routerProvider` from `@pankod/refine-nextjs-router`.

In order to define our resource action routes, we will make use of the `resources` array, and we will define our pages within the `pages` directory.

We will create a total of four pages for our resources, including:

- `pages/posts/index.tsx` - A list page for posts.
- `pages/posts/show/[id].tsx` - A detail page for posts.
- `pages/categories/index.tsx` - A list page for categories.
- `pages/categories/show/[id].tsx` - A detail page for categories.

Additionally, we will create a page for the index route, and use it to redirect to the posts resource:

- `pages/index.tsx` - An index page.

To begin setting up the **refine** app, we will first initialize it within the _app.tsx file.
```

```tsx title=_app.tsx
import { Refine } from "@pankod/refine-core";
import dataProvider from "@pankod/refine-simple-rest";
import routerProvider from "@pankod/refine-nextjs-router";

import { Layout } from "components/Layout";

function App({ Component, pageProps }: AppProps): JSX.Element {
    return (
        <Refine
            dataProvider={dataProvider("https://api.fake-rest.refine.dev")}
            // highlight-next-line
            routerProvider={routerProvider}
            resources={[
                {
                    name: "posts",
                    list: "/posts",
                    show: "/posts/show/:id",
                },
                {
                    name: "categories",
                    list: "/categories",
                    show: "/categories/show/:id",
                }
            ]}
        >
            <Layout>
                <Component {...pageProps} />
            </Layout>
        </Refine>
    );
}

```

:::tip

> Next.js uses the bracket syntax (`[param]`) for dynamic routes but **refine** uses the colon syntax (`:param`) for route parameters. This won't cause any problems since **refine** only uses the colon syntax as an indicator for route parameters and the communication between **refine** and the router is handled by the `routerProvider` prop.

Reviewer: ```suggestion
It's worth noting that Next.js uses the bracket syntax (`[param]`) for dynamic routes, while **refine** uses the colon syntax (`:param`) for route parameters.

Despite this difference in syntax, there won't be any issues when using **refine** with Next.js. This is because **refine** only uses the colon syntax as an indicator for route parameters, and the communication between **refine** and the router is handled by the `routerProvider` prop.
```

:::

:::note Additional Parameters and Nesting

> Your action definitions in the resources can contain additional parameters and nested routes. Passing these parameters when navigating to the pages are handled by the current available parameters and the `meta` props of the related hooks and components.

Reviewer: ```suggestion
When defining actions in your resources, you can include additional parameters and nested routes. Passing these parameters when navigating to the related pages is handled by the available parameters and the meta prop of the related hooks and components.
```

> **refine** supports route parameters defined with `:param` syntax. You can use these parameters in your action definitions and create your routes accordingly. For example, if you have a `posts` resource and you want to create a route for the `show` action of a specific post, you can define the `show` action as `/posts/show/:id` and use the `id` parameter in your component.

Reviewer: ```suggestion
It's worth noting that **refine** supports route parameters defined with the `:param` syntax. You can use these parameters in your action definitions and create your routes accordingly.

For example, if you have a `posts` resource and want to create a route for the `show` action of a specific post, you can define the show action as `/posts/show/:id` and use the id parameter in your component.
```

            {!isLoading && (
                <ul>
                    {tableData?.map((category) => (
                        <li key={category.id}>
                            <Link href={`/categories/show/${category.id}`}>
                                {category.label}
                            </Link>
                        </li>
                    ))}
                </ul>
            )}
        </div>
    )
}
```

```tsx title=pages/categories/show/[id].tsx
import { useShow } from "@pankod/refine-core";

type ICategory = {
    id: string;
    label: string;
}

export default function CategoryShow() {
    // `categories` resource and the `id` will be inferred from the route.
    // Because we've defined `/categories/show/:id` as the `show` action of the `categories` resource.
    const { queryResult: { data, isLoading } } = useShow<ICategory>();

    const categoryData = data?.data;

    return (
        <div>
            <h1>{categoryData?.label}</h1>
        </div>
    );
}
```

Now, we'll use [`NavigateToResource`](#navigatetoresource) component to redirect to the `posts` resource when the user visits the home page.

:::tip

> Even though we're using the `NavigateToResource` component, when using Next.js it's better to handle such redirect operations in the server side rather than the client side. You can use the `getServerSideProps` function to redirect the user to the `/posts`.

Reviewer: ```suggestion
Even though we are using the `NavigateToResource` component here, it's better to handle redirection operations on the server-side rather than the client-side when working with Next.js.

For this purpose, you can make use of the `getServerSideProps` function to redirect the user to the `/posts` resource.
```

:::

```tsx title=pages/index.tsx
import { NavigateToResource } from "@pankod/refine-nextjs-router";

export default function Home() {
    return (
        <NavigateToResource />
    );
}
```

### `/app` Directory

> We'll use the `routerProvider` from `@pankod/refine-nextjs-router/app` to set up the router bindings for **refine**. We'll define the action routes for our resources in the `resources` array and define our pages in the `app` directory.

Reviewer: ```suggestion
We will use the `routerProvider` from `@pankod/refine-nextjs-router/app` to establish the router bindings for **refine**.

The resources array will contain the action routes for our `resources`, and our pages will be defined in the `app` directory.
```

> We'll create four routes for our resources:

Reviewer: ```suggestion
To create routes for our `resources`, we will create the following routes:
```

- `app/posts/page.tsx` - List page for posts
- `app/posts/show/[id]/page.tsx` - Detail page for posts
- `app/categories/page.tsx` - List page for categories
- `app/categories/show/[id]/page.tsx` - Detail page for categories

> And we'll create one route for the index and use it to redirect to the `posts` resource:

Reviewer: ```suggestion
Additionally, we will create a route for the index page and use it to redirect to the `posts` resource:
```

    children,
}: {
    children: React.ReactNode;
}) {
    return (
        <html lang="en">
            <body>
                <Refine
                    dataProvider={dataProvider("https://api.fake-rest.refine.dev")}
                    // highlight-next-line
                    routerProvider={routerProvider}
                    resources={[
                        {
                            name: "posts",
                            list: "/posts",
                            show: "/posts/show/:id",
                        },
                        {
                            name: "categories",
                            list: "/categories",
                            show: "/categories/show/:id",
                        }
                    ]}
                >
                    {/* We're defining `Layout` here but you might want to have different layouts per your page. */}
                    {/* This is totally fine for refine, you can place your Layout wherever you like. */}
                    <Layout>
                        {children}
                    </Layout>
                </Refine>
            </body>
        </html>
    );
}

```

:::tip

> Next.js uses the bracket syntax (`[param]`) for dynamic routes but **refine** uses the colon syntax (`:param`) for route parameters. This won't cause any problems since **refine** only uses the colon syntax as an indicator for route parameters and the communication between **refine** and the router is handled by the `routerProvider` prop.

Reviewer: I think this part is duplicated.

}

export default function CategoryShow() {
    // `categories` resource and the `id` will be inferred from the route.
    // Because we've defined `/categories/show/:id` as the `show` action of the `categories` resource.
    const { queryResult: { data, isLoading } } = useShow<ICategory>();

    const categoryData = data?.data;

    return (
        <div>
            <h1>{categoryData?.label}</h1>
        </div>
    );
}
```

Now, we'll use [`NavigateToResource`](#navigatetoresource) component to redirect to the `posts` resource when the user visits the home page.

:::tip

Even though we're using the `NavigateToResource` component, when using Next.js it's better to handle such redirect operations in the server side rather than the client side. You can use the `getServerSideProps` function to redirect the user to the `/posts`.

:::

```tsx title=app/page.tsx
import { NavigateToResource } from "@pankod/refine-nextjs-router/app";

export default function IndexPage() {
    return (
        <NavigateToResource />
    );
}
```

## Additional Components

`@pankod/refine-nextjs-router` package also includes some additional components that can be useful in some cases.

### `RefineRoutes`

> This component can be used to render the matching route in your resources by using the `resources` prop. It will only take effect if the action properties in the resource definitions are assigned to components or objects with `component` property.

Reviewer: ```suggestion
This component allows you to render the matching route for your resources using the `resources` prop, but only if the action properties in the resource definitions are assigned to components or objects with `component` property. 
```

> It will render the component for the matching route and pass it as a `JSX.Element` to the `children` function. You can use this to render your components in a single catch-all route. If there's no matching route `undefined` will be passed to the `children` function. In this case, you can render an error page or redirect the user to another page.

Reviewer: ```suggestion
The component for the matching route will be rendered and passed to the children function as a `JSX.Element`.This allows you to render your components in a single catch-all route.

If there is no matching route, the `children` function will receive `undefined`. In this case, you can choose to render an error page or redirect the user to another page.
```

:::tip

> This is not the recommended way to create pages since it may limit the use of the full potential of your router. It is recommended to create your pages manually.

Reviewer: ```suggestion
It is not recommended to use this method to create pages, as it may limit the full potential of your router.

It is recommended to create your pages manually instead.
```
